src/base/memory_monitor_sr.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ *
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * memory_monitor_sr.cpp - implementation of memory monitoring manager
+ */
+
+#include <cstring>
+
+#include "perf.hpp"
+#include "error_manager.h"
+#include "memory_monitor_sr.hpp"
+
+#include "log_impl.h"
+
+namespace cubperf
+{
+  memory_monitor *mmon_Gl = nullptr;
+
+  const char *mmon_subcomponent::get_name ()
+  {
+    return m_subcomp_name.c_str ();
+  }
+
+  void mmon_subcomponent::add_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_add (size);
+  }
+
+  void mmon_subcomponent::sub_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_sub (size);
+  }
+
+  const char *mmon_component::get_name ()
+  {
+    return m_comp_name.c_str ();
+  }
+
+  void mmon_component::add_stat (uint64_t size, int subcomp_idx, bool init, bool expand)
```

Comment:
How about considering make an API function to inform chaning the state from init -> service?

In this approach(use init parameter), every time need to check init value even if it only necessary in the booting step. If you use API funtion, then you just call it at some point before service state.

For example,

init_stat = 0;
cur_stat = 0;

call add_stat () -> { cur_stat += size; }
call go_to_service_state () -> { init_stat = cur_stat; cur_stat = 0; }
call add_stat () -> { cur_stat += size; }

Reply:
About expand, I will leave a comment if it needed after checking the usage of it later.

Reply:
I think it's a good idea, but it seems that we need to consider the "appropriate" timing for when this function will be inserted.